### PR TITLE
[1848] implement receivership 

### PR DESCRIPTION
**Pull request by bentziaxl**
cert limit goes down based on number of companies in receivership and player count
president has further reduction for causing a company to go bankrupt
company pays compensation, if not enough president is liable. 
No personal bankruptcy, player has negative money, and must make it back by earning it elsewhere.

### CODE REVIEW

File: lib/engine/game/g_1848/game.rb [hunk 1]
```diff
@@ -740,6 +755,80 @@ def revenue_for(route, stops)
...
+            next if corporation.president?(player)
+
+            amount = player.num_shares_of(corporation) * per_share
+            next if amount.zero?
+
+            payouts[player] = amount
+            corporation.spend(amount, player, check_cash: false, borrow_from: corporation.owner)
+          end
+
+          if payouts.any?
```
Comment (tobymao) [inline lib/engine/game/g_1848/game.rb:785]: unless empty

File: lib/engine/game/g_1848/step/cash_crisis.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          def description
+            'Cash Crisis'
+          end
+
+          def cash_crisis?
+            true
+          end
+
+          def active?
+            active_entities.any?
```
Comment (tobymao) [inline lib/engine/game/g_1848/step/cash_crisis.rb:34]: !active_entities.empty?

File: lib/engine/game/g_1848/game.rb [hunk 3]
```diff
@@ -740,6 +755,80 @@ def revenue_for(route, stops)
...
         end
+
+        # recievership
+
+        def close_corporation(corporation, quiet: false)
+          @close_corp_count += 1
+          @player_corp_close_count[corporation.owner] += 1
+
+          # boe gets all the tokens
+          corporation.tokens.select(&:used).each do |token|
```
Comment (tobymao) [inline lib/engine/game/g_1848/game.rb:766]: remove the select and just do each do |token|
next used

File: lib/engine/game/g_1848/step/cash_crisis.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  module Game
+    module G1848
+      module Step
+        class CashCrisis < Engine::Step::Base
+          include Engine::Step::EmergencyMoney
+
+          def actions(entity)
+            return [] unless entity == current_entity
+
+            if @active_entity.nil?
```
Comment (tobymao) [inline lib/engine/game/g_1848/step/cash_crisis.rb:16]: unless @active_entity

File: lib/engine/game/g_1848/step/cash_crisis.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+          def process_sell_shares(action)
+            super
+            return if action.entity.cash.negative?
+
+            @active_entity = nil
+          end
+
+          def sellable_shares?(entity)
+            sum = entity.shares.sum do |share|
```
Comment (tobymao) [inline lib/engine/game/g_1848/step/cash_crisis.rb:62]: shares.count { can_sell?}

Comment (tobymao) [reply lib/engine/game/g_1848/step/cash_crisis.rb:62]: hm doesn't look like you even use sum, just do

entity.shares.any? | can_sell |

File: lib/engine/game/g_1848/game.rb (hunk 1 above)
Comment (bentziaxl) [reply lib/engine/game/g_1848/game.rb]: fixed. Btw why is unless empty better than if any?

File: lib/engine/game/g_1848/step/cash_crisis.rb (hunk 2 above)
Comment (bentziaxl) [reply lib/engine/game/g_1848/step/cash_crisis.rb]: fixed

File: lib/engine/game/g_1848/game.rb (hunk 3 above)
Comment (bentziaxl) [reply lib/engine/game/g_1848/game.rb]: fixe

File: lib/engine/game/g_1848/step/cash_crisis.rb (hunk 4 above)
Comment (bentziaxl) [reply lib/engine/game/g_1848/step/cash_crisis.rb]: fixed

File: lib/engine/game/g_1848/step/cash_crisis.rb (hunk 5 above)
Comment (bentziaxl) [reply lib/engine/game/g_1848/step/cash_crisis.rb]: fixed

File: lib/engine/game/g_1848/game.rb (hunk 1 above)
Comment (tobymao) [reply lib/engine/game/g_1848/game.rb]: any checks nulls so it's less efficient than empty which is just a size check

**Pull request merged**